parser/transaction.go

```diff
@@ -397,76 +381,193 @@ func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+		return nil, errors.New("could not read nVersionGroupId")
+	}
+	if !s.ReadUint32(&tx.nLockTime) {
+		return nil, errors.New("could not read nLockTime")
+	}
+	if !s.ReadUint32(&tx.nExpiryHeight) {
+		return nil, errors.New("could not read nExpiryHeight")
+	}
+	s, err = tx.ParseTransparent([]byte(s))
+	if err != nil {
+		return nil, err
+	}
 
-	if tx.version >= 4 && (spendCount+outputCount > 0) {
-		if !s.ReadBytes(&tx.bindingSig, 64) {
-			return nil, errors.New("could not read bindingSig")
+	var spendCount, outputCount int
+	if !s.ReadCompactSize(&spendCount) {
+		return nil, errors.New("could not read nShieldedSpend")
+	}
+	tx.shieldedSpends = make([]*spend, spendCount)
+	for i := 0; i < spendCount; i++ {
+		newSpend := &spend{}
+		s, err = newSpend.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Spend")
+		}
+		tx.shieldedSpends[i] = newSpend
+	}
+	if !s.ReadCompactSize(&outputCount) {
+		return nil, errors.New("could not read nShieldedOutput")
+	}
+	tx.shieldedOutputs = make([]*output, outputCount)
+	for i := 0; i < outputCount; i++ {
+		newOutput := &output{}
+		s, err = newOutput.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Output")
+		}
+		tx.shieldedOutputs[i] = newOutput
+	}
+	if spendCount+outputCount > 0 && !s.ReadInt64(&tx.valueBalanceSapling) {
+		return nil, errors.New("could not read valueBalance")
+	}
+	if spendCount > 0 && !s.Skip(32) {
+		return nil, errors.New("could not skip anchorSapling")
+	}
+	if !s.Skip(192 * spendCount) {
+		return nil, errors.New("could not skip vSpendProofsSapling")
+	}
+	if !s.Skip(64 * spendCount) {
+		return nil, errors.New("could not skip vSpendAuthSigsSapling")
+	}
+	if !s.Skip(192 * outputCount) {
+		return nil, errors.New("could not skip vOutputProofsSapling")
+	}
+	if spendCount+outputCount > 0 && !s.ReadBytes(&tx.bindingSigSapling, 64) {
+		return nil, errors.New("could not read bindingSigSapling")
+	}
+	var actionsCount, proofsCount int
+	if !s.ReadCompactSize(&actionsCount) {
+		return nil, errors.New("could not read nActionsOrchard")
+	}
+	if !s.Skip(820 * actionsCount) {
+		return nil, errors.New("could not skip vActionsOrchard")
+	}
+	if actionsCount > 0 {
+		if !s.Skip(1) {
+			return nil, errors.New("could not skip flagsOrchard")
+		}
+		if !s.Skip(8) {
+			return nil, errors.New("could not skip valueBalanceOrchard")
+		}
+		if !s.Skip(32) {
+			return nil, errors.New("could not skip anchorOrchard")
+		}
+		if !s.ReadCompactSize(&proofsCount) {
+			return nil, errors.New("could not read sizeProofsOrchard")
+		}
+		if !s.Skip(proofsCount) {
+			return nil, errors.New("could not skip proofsOrchard")
+		}
+		if !s.Skip(64 * actionsCount) {
+			return nil, errors.New("could not skip vSpendAuthSigsOrchard")
+		}
+		if !s.Skip(64) {
+			return nil, errors.New("could not skip bindingSigOrchard")
 		}
 	}
+	return s, nil
```

Comment:
Should there be a check here to validate that `s` is empty? I'm not sure why these are implemented as methods on a mutable `Transaction` object, rather than as functions that return a `Transaction`. The way this is written, the transaction is not guaranteed to be in an empty state before you start parsing.

In general (contrary to how the C++ codebase does it) parsers should be functions of the form `byte[] -> type | error`; in Go this is hard because you don't have sum types, so you have to return `type & error` instead and check that the error is nil, but that still allows our parsers to be functions rather than mutating methods.

Reply:
Yeah, I agree that the larger refactoring isn't something that should be done as part of this PR.